File: sdk/src/privportal_sdk/middleware.py
"""Core sanitize/resolve middleware.

Loads identity mappings from PolarPrivate once, then performs
pure in-memory string replacement for every message.

PolarPrivate sanitize/mappings API requires no authentication.
"""

from __future__ import annotations

import re
import threading
from dataclasses import dataclass
from typing import Any

import httpx


@dataclass
class _IdentityEntry:
    key: str
    value: str
    project_id: str | None
    placeholder: str = ""

    def __post_init__(self) -> None:
        self.placeholder = f"[[{self.key}]]"


@dataclass
class _SecretEntry:
    key: str
    project_id: str | None
    placeholder: str = ""

    def __post_init__(self) -> None:
        self.placeholder = f"[[secret_ref.{self.key}]]"


class PrivPortalMiddleware:
    """Stateful middleware that sanitizes and resolves text using PolarPrivate mappings.

    PolarPrivate sanitize/mappings API requires no authentication —
    any localhost caller can use it without tokens or cookies.

    Usage::

        from privportal_sdk import PrivPortalMiddleware

        mw = PrivPortalMiddleware("http://127.0.0.1:12790")
        mw.load_mappings()
    """

    def __init__(
        self,
        base_url: str = "http://127.0.0.1:12790",
        project_id: str | None = None,
        *,
        auto_load: bool = False,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._project_id = project_id
        self._identities: list[_IdentityEntry] = []
        self._secrets: list[_SecretEntry] = []
        self._value_to_placeholder: dict[str, str] = {}
        self._placeholder_to_value: dict[str, str] = {}
        self._sanitize_pattern: re.Pattern[str] | None = None
        self._lock = threading.Lock()
        self._loaded = False

        if auto_load:
            self.load_mappings()

    @property
    def is_loaded(self) -> bool:
        return self._loaded

    def load_mappings(self, timeout: float = 5.0) -> None:
        """Fetch mappings from PolarPrivate API and build lookup tables."""
        params: dict[str, Any] = {}
        if self._project_id is not None:
            params["project_id"] = self._project_id

        with httpx.Client(timeout=timeout) as client:
            resp = client.get(
                f"{self._base_url}/api/sanitize/mappings",
                params=params,
            )
            resp.raise_for_status()
            data = resp.json()
        self._rebuild(data)

    async def aload_mappings(self, timeout: float = 5.0) -> None:
        """Async variant of load_mappings."""
        params: dict[str, Any] = {}
        if self._project_id is not None:
            params["project_id"] = self._project_id

        async with httpx.AsyncClient(timeout=timeout) as client:
            resp = await client.get(
                f"{self._base_url}/api/sanitize/mappings",
                params=params,
            )
            resp.raise_for_status()
            data = resp.json()
        self._rebuild(data)

    def load_from_dict(self, data: dict[str, Any]) -> None:
        """Load mappings from a raw dict (useful for testing without a running server)."""
        self._rebuild(data)

    def _rebuild(self, data: dict[str, Any]) -> None:
        """Rebuild internal lookup tables from API response."""
        with self._lock:
            self._identities = [
                _IdentityEntry(
                    key=item["key"],
                    value=item["value"],
                    project_id=item.get("project_id"),
                )
                for item in data.get("identities", [])
            ]
            self._secrets = [
                _SecretEntry(
                    key=item["key"],
                    project_id=item.get("project_id"),
                )
                for item in data.get("secrets", [])
            ]

            self._value_to_placeholder = {}
            self._placeholder_to_value = {}

            # Sort by value length descending so longer values match first
            # (e.g. "张三丰" matches before "张三")
            sorted_ids = sorted(self._identities, key=lambda e: len(e.value), reverse=True)
            for entry in sorted_ids:
                if entry.value:
                    self._value_to_placeholder[entry.value] = entry.placeholder
                    self._placeholder_to_value[entry.placeholder] = entry.value

            if self._value_to_placeholder:
                escaped = [re.escape(v) for v in self._value_to_placeholder]
                self._sanitize_pattern = re.compile("|".join(escaped))
            else:
                self._sanitize_pattern = None

            self._loaded = True

    def sanitize(self, text: str) -> str:
        """Replace known identity values with their placeholders.

        Returns the text with all recognized PII replaced by [[key]] tokens.
        This is the INBOUND path (user message → LLM).
        """
        if not self._loaded or self._sanitize_pattern is None:
            return text

        def _replace(match: re.Match[str]) -> str:
            return self._value_to_placeholder[match.group(0)]

        return self._sanitize_pattern.sub(_replace, text)

    def resolve(self, text: str) -> str:
        """Replace placeholders with real identity values.

        Returns the text with [[key]] tokens replaced by actual values.
        This is the OUTBOUND path (LLM reply → user).
        """
        if not self._loaded or not self._placeholder_to_value:
            return text

        result = text
        for placeholder, value in self._placeholder_to_value.items():
            result = result.replace(placeholder, value)
        return result

    def detect_leaks(self, text: str) -> list[dict[str, str]]:
        """Check if text contains any known identity values (leak detection).

        Returns a list of detected leaks with key and matched value.
        Useful for auditing LLM outputs before sending to user.
        """
        if not self._loaded or self._sanitize_pattern is None:
            return []

        leaks: list[dict[str, str]] = []
        for match in self._sanitize_pattern.finditer(text):
            value = match.group(0)
            placeholder = self._value_to_placeholder[value]
            leaks.append({
                "key": placeholder.strip("[]"),
                "value": value,
                "position": match.start(),
            })
        return leaks

    @property
    def identity_count(self) -> int:
        return len(self._identities)

    @property
    def secret_count(self) -> int:
        return len(self._secrets)

    def __repr__(self) -> str:
        return (
            f"PrivPortalMiddleware(base_url={self._base_url!r}, "
            f"loaded={self._loaded}, "
            f"identities={self.identity_count}, "
            f"secrets={self.secret_count})"
        )

File: sdk-ts/src/middleware.ts
/**
 * PolarPrivate TypeScript SDK — sanitize/resolve middleware.
 *
 * Mirrors the Python `privportal-sdk` API. Loads identity mappings from
 * PolarPrivate once, then performs pure in-memory string replacement.
 *
 * @example
 * ```ts
 * const mw = new PrivPortalMiddleware("http://127.0.0.1:12790");
 * await mw.loadMappings();
 * const safe = mw.sanitize("你好，我是张三");     // → "你好，我是[[identity.student.name]]"
 * const real = mw.resolve("[[identity.student.name]] 你好"); // → "张三你好"
 * ```
 */

export interface IdentityMapping {
  key: string;
  value: string;
  project_id: string | null;
}

export interface SecretMapping {
  key: string;
  project_id: string | null;
}

export interface MappingsResponse {
  identities: IdentityMapping[];
  secrets: SecretMapping[];
  version: string;
}

export interface LeakInfo {
  key: string;
  value: string;
  position: number;
}

export interface PrivPortalOptions {
  baseUrl?: string;
  projectId?: string | null;
  autoLoad?: boolean;
}

function escapeRegex(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

export class PrivPortalMiddleware {
  private _baseUrl: string;
  private _projectId: string | null;
  private _identities: IdentityMapping[] = [];
  private _secrets: SecretMapping[] = [];
  private _valueToPlaceholder = new Map<string, string>();
  private _placeholderToValue = new Map<string, string>();
  private _sanitizePattern: RegExp | null = null;
  private _loaded = false;

  constructor(options: PrivPortalOptions | string = {}) {
    if (typeof options === "string") {
      options = { baseUrl: options };
    }
    this._baseUrl = (options.baseUrl ?? "http://127.0.0.1:12790").replace(
      /\/+$/,
      ""
    );
    this._projectId = options.projectId ?? null;
  }

  get isLoaded(): boolean {
    return this._loaded;
  }

  get identityCount(): number {
    return this._identities.length;
  }

  get secretCount(): number {
    return this._secrets.length;
  }

  /**
   * Fetch mappings from PolarPrivate API and build lookup tables.
   * The /api/sanitize/mappings endpoint requires no authentication
   * (any localhost caller can use it).
   */
  async loadMappings(timeout = 5000): Promise<void> {
    const url = new URL(`${this._baseUrl}/api/sanitize/mappings`);
    if (this._projectId) {
      url.searchParams.set("project_id", this._projectId);
    }

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeout);
    try {
      const resp = await fetch(url.toString(), {
        signal: controller.signal,
      });
      if (!resp.ok) {
        throw new Error(
          `PolarPrivate API returned ${resp.status}: ${resp.statusText}`
        );
      }
      const data = (await resp.json()) as MappingsResponse;
      this._rebuild(data);
    } finally {
      clearTimeout(timer);
    }
  }

  /**
   * Load mappings from a raw object (useful for testing without a running server).
   */
  loadFromObject(data: MappingsResponse): void {
    this._rebuild(data);
  }

  private _rebuild(data: MappingsResponse): void {
    this._identities = data.identities ?? [];
    this._secrets = data.secrets ?? [];

    this._valueToPlaceholder.clear();
    this._placeholderToValue.clear();

    const sorted = [...this._identities].sort(
      (a, b) => b.value.length - a.value.length
    );

    for (const entry of sorted) {
      if (!entry.value) continue;
      const placeholder = `[[${entry.key}]]`;
      this._valueToPlaceholder.set(entry.value, placeholder);
      this._placeholderToValue.set(placeholder, entry.value);
    }

    if (this._valueToPlaceholder.size > 0) {
      const escaped = [...this._valueToPlaceholder.keys()].map(escapeRegex);
      this._sanitizePattern = new RegExp(escaped.join("|"), "g");
    } else {
      this._sanitizePattern = null;
    }

    this._loaded = true;
  }

  /**
   * Replace known identity values with their placeholders.
   * This is the INBOUND path (user message → LLM).
   */
  sanitize(text: string): string {
    if (!this._loaded || !this._sanitizePattern) return text;
    this._sanitizePattern.lastIndex = 0;
    return text.replace(this._sanitizePattern, (match) => {
      return this._valueToPlaceholder.get(match) ?? match;
    });
  }

  /**
   * Replace placeholders with real identity values.
   * This is the OUTBOUND path (LLM reply → user).
   */
  resolve(text: string): string {
    if (!this._loaded || this._placeholderToValue.size === 0) return text;
    let result = text;
    for (const [placeholder, value] of this._placeholderToValue) {
      result = result.replaceAll(placeholder, value);
    }
    return result;
  }

  /**
   * Check if text contains any known identity values (leak detection).
   * Returns a list of detected leaks with key and matched value.
   */
  detectLeaks(text: string): LeakInfo[] {
    if (!this._loaded || !this._sanitizePattern) return [];
    this._sanitizePattern.lastIndex = 0;
    const leaks: LeakInfo[] = [];
    let match: RegExpExecArray | null;
    while ((match = this._sanitizePattern.exec(text)) !== null) {
      const value = match[0];
      const placeholder = this._valueToPlaceholder.get(value);
      if (placeholder) {
        leaks.push({
          key: placeholder.slice(2, -2),
          value,
          position: match.index,
        });
      }
    }
    return leaks;
  }

  toString(): string {
    return `PrivPortalMiddleware(baseUrl=${this._baseUrl}, loaded=${this._loaded}, identities=${this.identityCount}, secrets=${this.secretCount})`;
  }
}
